// noform/lib/validate/validate.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _messages = require('./messages');

var _messages2 = _interopRequireDefault(_messages);

var _rules = require('./rules');

var _rules2 = _interopRequireDefault(_rules);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Validate = function () {
    function Validate(schema) {
        var locale = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'zh_CN';

        _classCallCheck(this, Validate);

        this.schema = schema;
        this.setLocal(locale);
        this.keys = Object.keys(this.schema);
    }

    Validate.prototype.setLocal = function setLocal(locale) {
        this.messages = (0, _messages2['default'])(locale);
    };

    Validate.prototype.test = function test(value) {
        var _this = this;

        return this.keys.reduce(function (s, key) {
            return s && new _rules2['default'](_this.schema[key]).test(value[key]);
        }, true);
    };

    Validate.prototype.validate = function validate(value) {
        var _this2 = this;

        var ret = {};
        var hasPromise = false;
        var validatePromise = this.keys.map(function (key) {
            var result = new _rules2['default'](_this2.schema[key]).validate(value[key], key, value);
            if (result instanceof Promise) {
                hasPromise = true;
                return result.then(function (res) {
                    if (res) {
                        ret[key] = res;
                    }
                });
            }
            if (result) {
                ret[key] = result;
            }
            return null;
        });
        if (hasPromise) {
            return Promise.all(validatePromise).then(function () {
                if (Object.keys(ret).length === 0) {
                    return null;
                }
                return ret;
            });
        }
        if (Object.keys(ret).length === 0) {
            return null;
        }
        return ret;
    };

    return Validate;
}();

exports['default'] = Validate;
module.exports = exports['default'];
