// noform/lib/dialog/next.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _DialogFormFactory = require('./DialogFormFactory');

var _DialogFormFactory2 = _interopRequireDefault(_DialogFormFactory);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var DialogFormWrapper = function DialogFormWrapper(NextSource) {
    var Dialog = NextSource.Dialog,
        Button = NextSource.Button;

    Dialog.show = Dialog.confirm;

    var compatiMap = {
        show: function show(options) {
            var width = options.width,
                className = options.className;


            return _extends({}, options, {
                width: width,
                className: (className || '') + ' dialog-form-wrapper',
                needWrapper: false
            });
        },
        dialogInstance: function dialogInstance(_dialogInstance) {
            return _dialogInstance;
        },
        btnLoadingProps: 'loading'
    };

    return new _DialogFormFactory2['default']({
        Dialog: Dialog,
        Button: Button,
        compatiMap: compatiMap
    });
};

exports['default'] = DialogFormWrapper;
module.exports = exports['default'];
