// noform/lib/static/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
// 组件变更
var ANY_CHANGE = 'ANY_CHANGE';

// 组件属性
var CHANGE = 'CHANGE';
var FOCUS = 'FOCUS';
var BLUR = 'BLUR';

// 基础属性变更
var VALUE_CHANGE = 'VALUE_CHANGE';
var ERROR_CHANGE = 'ERROR_CHANGE';
var PROPS_CHANGE = 'PROPS_CHANGE';
var STATUS_CHANGE = 'STATUS_CHANGE';
var BASIC_EVENT = {
    value: VALUE_CHANGE,
    error: ERROR_CHANGE,
    props: PROPS_CHANGE,
    status: STATUS_CHANGE
};

// 状态枚举
var EDIT = 'edit';
var PREVIEW = 'preview';
var DISABLED = 'disabled';
var HIDDEN = 'hidden';
var STATUS_ENUMS = new Set([EDIT, PREVIEW, DISABLED, HIDDEN]);

exports.ANY_CHANGE = ANY_CHANGE;
exports.BASIC_EVENT = BASIC_EVENT;
exports.CHANGE = CHANGE;
exports.BLUR = BLUR;
exports.FOCUS = FOCUS;
exports.VALUE_CHANGE = VALUE_CHANGE;
exports.ERROR_CHANGE = ERROR_CHANGE;
exports.PROPS_CHANGE = PROPS_CHANGE;
exports.STATUS_CHANGE = STATUS_CHANGE;
exports.STATUS_ENUMS = STATUS_ENUMS;
exports.EDIT = EDIT;
exports.PREVIEW = PREVIEW;
exports.DISABLED = DISABLED;
exports.HIDDEN = HIDDEN;
